### PR TITLE
implement "engine" for managing inference/codegen

### LINKED ISSUE #53433: regression in thread-safety with data-race on mi->inInference field
The changes in #53219 appears to have introduced a datarace on the `mi->inInference` field, which seems likely to lead to incorrect compilation and bad performance of threaded code. It was previously protected by the codegen lock.

_Originally posted by @vtjnash in https://github.com/JuliaLang/julia/pull/53219#discussion_r1495076456_

Posting as an issue since the PR is already merged, and this is a regression in thread-safety

### LINKED ISSUE #53680: `llvmcall` requires the compiler
I'm recently seeing this error when using `TiffImages.jl` (dev branch) with multi-threading enabled

```julia
julia> VERSION
v"1.12.0-DEV.149"

julia> using TiffImages

julia> @time TiffImages.load("image.tif");
...
    nested task error: `llvmcall` requires the compiler
    Stacktrace:
     [1] macro expansion
       @ ~/.julia/packages/SIMD/2fAdM/src/LLVM_intrinsics.jl:673 [inlined]
     [2] shufflevector(x::NTuple{…}, ::Val{…})
       @ SIMD.Intrinsics ~/.julia/packages/SIMD/2fAdM/src/LLVM_intrinsics.jl:664
...

julia> @time TiffImages.load("image.tif");
  0.283151 seconds (164.76 k allocations: 282.782 MiB, 17.25% gc time, 38.50% compilation time)
```
Sometimes it works on the first try, but it always works on the second try

I don't get this with `v1.10.2`

### PR DESCRIPTION
**Pull request by vtjnash**
Continuing from previous PRs to making CodeInstance the primary means of tracking compilation, this introduces an "engine" which keeps track externally of whether a particular inference result is in progress and where. At present, this handles unexpected cycles by permitting both threads to work on it. This is likely to be optimal most of the time currently, until we have the ability to do work-stealing of the results.

To assist with that, CodeInstance is now primarily allocated by `jl_engine_reserve`, which also tracks that this is being currently inferred. This creates a sort of per-(MI,owner) tuple lock mechanism, which can be used with the double-check pattern to see if inference was completed while waiting on that. The `world` value is not included since that is inferred later, so there is a possibility that a thread waits only to discover that the result was already invalid before it could use it (though this should be unlikely).

The process then can notify when it has finished and wants to release the reservation lock on that identity pair. When doing so, it may also provide source code, allowing the process to potentially begin a threadpool to compile that result while the main thread continues with the job of inference.

Includes fix for #53434, by ensuring SOURCE_MODE_ABI results in the item going into the global cache.

Fixes #53433, as inInference is computed by the engine and protected by a lock, which also fixes #53680.

~~(n.b. somehow the finalizers_inhibited counter appears to be getting corrupted by the worlds and staged tests, so maybe it will be necessarily to shift exactly how that is done, if we do want to prevent finalizers from occurring here, e.g. `JULIA_CPU_THREADS=1 ../julia -p1 runtests.jl worlds int` causes an assertion failure. I suspect this is because it is Task-local state, but if something fails and we forget to reject it, then it will be subtracted from ptls2 instead, which might not be the same.)~~

### CODE REVIEW

File: src/engine.cpp [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+                break; // no cycle, since it is running
+            auto record2 = Reservations.find(key2);
+            if (record2 == Reservations.end())
+                break; // no cycle, since it is about to resume
+            assert(wait_tid != record2->second.tid);
+            wait_tid = record2->second.tid;
+        }
+        int8_t gc_state = jl_gc_safe_enter(ct->ptls);
+        Awaiting[tid] = key;
+        engine_wait.wait(lock);
```
Comment (gbaraldi) [inline src/engine.cpp:101]: Is there any benefit of making this scheduler aware?

File: src/engine.cpp [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+jl_code_instance_t *jl_engine_reserve(jl_method_instance_t *m, jl_value_t *owner)
+{
+    jl_task_t *ct = jl_current_task;
+    ct->ptls->engine_nqueued++; // disables finalizers until inference is finished on this method graph
```
Comment (gbaraldi) [inline src/engine.cpp:63]: Why is this necessary?

Comment (oscardssmith) [reply src/engine.cpp:63]: you don't want finalizers to run during inference because running finalizers might try to re-enter inference.

**Comment by gbaraldi** [comment]
Do we want to add something to the devdocs like:
The life of a codeinstance, all the way from someone requesting a MI to it being codegened?

File: src/engine.cpp (hunk 2 above)
Comment (gbaraldi) [reply src/engine.cpp:63]: Oh, ok. We really should do the finalizer thread thing, it probably solves quite a few of these issues

**Comment by Keno** [comment]
Overall, I think I like the direction. I think the only conceptual thing I flagged is that this start assuming that `owner` is an egal-comparable cache-key. In further development #54373 I ended up finding it useful to separate the egal-comparable key data and the non-volatile data. That said, I'm happy to keep going with this and figure out how to make that distinction later. I think in currently implemented code, it doesn't really matter.

**Comment by vtjnash** [comment]
> this start assuming that owner is an egal-comparable cache-key

I speculate that what will be happening here (eventually, if we need more flexibility), is that you would use a particular egal-comparable key for the engine reservation, but then if you later infer that you you want to refine that, the code would be allowed to do that, as long as that follows conceptually the same sort narrowing and lifecycle management that applies to the other fields. It is potentially very similar to how we deal with pkg precompile locks for example, where the pidfile is simply the package name, rather than being a guarantee that it is precisely the final content hash, and then we move it into place later.

**Comment by vtjnash** [comment]
Oh, also, of secondary note is that this is permitted to be reimplemented in Julia for anything that won't go through the JIT. The only reason this needs to be in C for the stdlib is that it needs to manage the state of things in the JIT before Julia is compiled and that adds some fairly strict concurrency constraints. For stuff that is merely reusing the inference machinery from Julia (e.g. not causing recursion through the JIT), it is permitted to reimplement this in Julia, in any similar way, using any alternate key comparator strategies.

**Comment by Keno** [comment]
Yeah, let's try it. As I said, directionally correct, so let's get the implementation in and then we can figure out what tweaks may be required for the new features that we want.

**Review by Keno — APPROVED** [review]

**Comment by vtjnash** [comment]
Ended up just using `~/julia/julia --project=.. bughunt.jl 'https://buildkite.com/julialang/julia-master/builds/37504#019032a2-da34-49fd-8b24-295852b1c908` and that pretty quickly let me identify the compiler bug (technically a C11 standards bug) in the version of gcc being used that was causing this to fail
